odl/tomo/backends/astra_setup.py

```diff
@@ -574,88 +588,90 @@ def astra_data(astra_geom, datatype, data=None, ndim=2, allow_copy=False):
-
-    # Mapping from interpolation type and geometry to ASTRA projector type.
-    # "I" means probably mathematically inconsistent. Some projectors are
-    # not implemented, e.g. CPU 3d projectors in general.
-    type_map_cpu = {'parallel': {'nearest': 'line',
-                                 'linear': 'linear'},  # I
-                    'fanflat': {'nearest': 'line_fanflat',
-                                'linear': 'line_fanflat'},  # I
-                    'parallel3d': {'nearest': 'linear3d',  # I
-                                   'linear': 'linear3d'},  # I
-                    'cone': {'nearest': 'linearcone',  # I
-                             'linear': 'linearcone'}}  # I
-    type_map_cpu['fanflat_vec'] = type_map_cpu['fanflat']
-    type_map_cpu['parallel3d_vec'] = type_map_cpu['parallel3d']
-    type_map_cpu['cone_vec'] = type_map_cpu['cone']
-
-    # GPU algorithms not necessarily require a projector, but will in future
-    # releases making the interface more coherent regarding CPU and GPU
-    type_map_cuda = {'parallel': 'cuda',  # I
-                     'parallel3d': 'cuda3d'}  # I
-    type_map_cuda['fanflat'] = type_map_cuda['parallel']
-    type_map_cuda['fanflat_vec'] = type_map_cuda['fanflat']
-    type_map_cuda['cone'] = type_map_cuda['parallel3d']
-    type_map_cuda['parallel3d_vec'] = type_map_cuda['parallel3d']
-    type_map_cuda['cone_vec'] = type_map_cuda['cone']
-
-    # create config dict
+    astra_geom = astra_proj_geom['type']
+    if (
+        astra_geom == 'parallel_vec'
+        and not astra_supports('par2d_vec_geometry')
+    ):
+        raise ValueError(
+            "'parallel_vec' geometry not supported in the current version "
+            "of ASTRA"
+        )
+
+    # Check if projector types are valid. We should not have to do this,
+    # but the errors from ASTRA are rather unspecific, so we check ourselves
+    # to know what's wrong.
+    astra_proj_type = str(astra_proj_type).lower()
+
+    if (
+        astra_proj_type == 'distance_driven'
+        and not astra_supports('par2d_distance_driven_proj')
+    ):
+        raise ValueError(
+            "'distance_driven' projector not supported in the current "
+            "version of ASTRA"
+        )
+
+    if astra_geom in {'parallel', 'parallel_vec'}:
+        valid_proj_types = ['line', 'linear', 'strip', 'cuda']
+        if astra_supports('par2d_distance_driven_proj'):
+            valid_proj_types.append('distance_driven')
+    elif astra_geom in {'fanflat', 'fanflat_vec'}:
+        valid_proj_types = ['line_fanflat', 'strip_fanflat', 'cuda']
+    elif astra_geom in {'parallel3d', 'parallel3d_vec'}:
+        valid_proj_types = ['linear3d', 'cuda3d']
+    elif astra_geom in {'cone', 'cone_vec'}:
+        valid_proj_types = ['linearcone', 'cuda3d']
+    else:
+        raise ValueError('invalid geometry type {!r}'.format(astra_geom))
+
+    if astra_proj_type not in valid_proj_types:
```

Comment:
How could this even happen given the above?

Reply:
I haven't checked this value up to that point yet, only collected the valid options.